[Tan Guan Qun] iP

monobeartae opened the pull request:
# MonoBot
> a CLI task scheduling bot developed by monobeartae

## Why you Should Look At My Code
Do you enjoy
- ~~pain~~
- ~~suffering~~
- coding standards
- java
- alignment issues?    

then looking at this project is for you! take a look at the code and **carefully** read through *everything*, as if you were the one who wrote it! so that you fully understand it!

## Using the bot
Don't forget to
1. Ensure you are using the right version of java jdk (even if it means having 3-4 java versions on your system because gradle doesnt support >19 😍 )
2. Bring eyedrops bcos CLI GUI
3. Smile :DDDDD   

With the bot, you can track how many times you
- [ ] cried
- [ ] slept (if any)
- [ ] cried even harder      

for **FREE**?!? wow sugoi! :oOoo ^-^

## Thank You
Hurry and Run
```
java -jar mono.jar
``` 
now!!!! jk i didnt even upload it

## By the Way
**Wishlist [Sedap!](https://store.steampowered.com/app/2650730/SEDAP_A_Culinary_Adventure/) on Steam NOW**. *(I contributed to it for 5 weeks over the winter break, it's a great game I promise, superior to overcooked, demo is free)* It's a 2-player co-op cooking + combat adventure so one person cooks the other whacks monsters for ingredients very fun no joke go grab ur boyfriend or girlfriend or friend or family to play and CLICK THE FAT ASS WISHLIST BUTTON

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

monobeartae committed:
Rename Bot, Greet and Exit

monobeartae committed:
Echo + Add A bit of Formatting

monobeartae committed:
Add, List

monobeartae committed:
Update .gitignore to ignore .class files

monobeartae committed:
Mark As Done

monobeartae committed:
Created Event, Todo, Deadline subclasses

monobeartae committed:
Track Event, Todo and Deadline

monobeartae committed:
Handle Errors

monobeartae committed:
Delete

monobeartae committed:
Automated Test UI

monobeartae committed:
Save Tasks Data

monobeartae committed:
Merge branch 'branch-Level-7'

monobeartae committed:
Use DateTime instead of String

monobeartae committed:
Close Scanner after use (I forgo oops)

monobeartae committed:
Merge branch 'branch-Level-8'

monobeartae committed:
Package classes + Update gitignore to ignore vscode files

monobeartae committed:
Add Find Feature

monobeartae committed:
Add Javadoc for all Tasks, SaveHandler, UserInputHandler, DateTime classes

monobeartae committed:
Refactor for OOP, separate UI, input parser and tasks logic

monobeartae committed:
Bug Fix: Missing Event Reference

monobeartae committed:
Merge branch 'master' into branch-A-JavaDoc

monobeartae committed:
Add JavaDoc Comments for MonoBot classes

monobeartae committed:
Merge branch 'master' into branch-Level-9

monobeartae committed:
Add Find Command Feature

monobeartae committed:
Move Main into app/ package

monobeartae committed:
Indent Switch case

monobeartae committed:
Merge branch 'branch-A-JavaDoc'

monobeartae committed:
Merge branch 'branch-A-CodingStandard'

monobeartae committed:
Merge branch 'branch-Level-9'

monobeartae committed:
Merge remote-tracking branch 'origin/add-gradle-support'

monobeartae committed:
Configure Gradle

monobeartae committed:
Major rework: shift errors to exceptions and decouple parser from bot

monobeartae committed:
Follow coding standards (i was blind)

monobeartae committed:
Add JUnit Tests for MonoBotInputParser

monobeartae committed:
Update gitignore to ignore .jar

monobeartae committed:
Add checkstyle to gradle build

monobeartae committed:
Remove text ui test folder

monobeartae committed:
Update README.md

monobeartae committed:
Use JavaFX GUI

monobeartae committed:
Merge branch 'branch-Level-10'

monobeartae committed:
Use Varargs

monobeartae committed:
Merge branch 'branch-A-Varargs'

SAN-MUYUN commented on src/main/java/app/SaveHandler.java:
```diff
@@ -0,0 +1,64 @@
+        ArrayList<Task> tasks = new ArrayList<>();
+        while (sc.hasNextLine()) {
+            String line = sc.nextLine();
+            tasks.add(Task.decodeTask(line));
```
I like how clean this code is!

SAN-MUYUN commented on src/main/java/app/commands/Command.java:
```diff
@@ -0,0 +1,21 @@
+package app.commands;
+
```
Perhaps you should add a header comment here?

monobeartae replied:
Good point, will add it in, thank you!

SAN-MUYUN commented on src/main/java/app/commands/StringCommand.java:
```diff
@@ -0,0 +1,15 @@
+public class StringCommand extends Command {
+    private String keyword;
+
+    public StringCommand(CommandType type, String s) {
```
perhaps a more meaningful variable name can be used instead of ``` s ```?

SAN-MUYUN commented on src/main/java/app/monobot/MonoBot.java:
```diff
@@ -0,0 +1,180 @@
+                throw new CommandTypeMismatchException(type, TaskIndexCommand.class, cmd.getClass());
+            }
+            this.unmarkCompletedTask(((TaskIndexCommand) cmd).getIndex());
+            break;
```
very clean piece of code here!

SAN-MUYUN commented on src/main/java/app/monobot/MonoBot.java:
```diff
@@ -0,0 +1,180 @@
+     * @param cmd Command
+     * @throws MonoBotException
+     */
+    public void processCommand(Command cmd) throws MonoBotException {
```
would it be better if more specific types of Exceptions are thrown? 🤔

monobeartae replied:
As my other exceptions inherit from MonoBotException, it is just a base class so that I can throw all kinds of more specific exceptions in the nested function calls! I do think I did not create a subclass for certain errors tho, I got lazy, will make those changes maybe! thanks for pointing this out!

SAN-MUYUN commented on src/main/java/app/tasks/Event.java:
```diff
@@ -0,0 +1,48 @@
+            e.markAsComplete();
+        }
+        return e;
+    }
```
I like your design of code here! 👍

SAN-MUYUN reviewed (commented):
Some nits to be fixed, but overall LGTM! 👍

monobeartae committed:
Merge branch 'master' of https://github.com/monobeartae/ip

monobeartae committed:
Add Header JavaDoc comment for Command class

monobeartae committed:
Add minor fix

Jensen188 commented on src/main/java/app/SaveHandler.java:
```diff
@@ -0,0 +1,64 @@
+        try {
+            if (!saveFile.exists()) {
+                saveFile.createNewFile();
+            }
```
Is it better to use `File.mkdirs()` [File.mkdirs()](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/File.html#mkdirs()) since app/data directory might not exist when running the app at the first time?

Jensen188 commented on src/main/java/app/SaveHandler.java:
```diff
@@ -0,0 +1,64 @@
+ * Class to handle saving and loading tasks data for MonoBot
+ */
+public class SaveHandler {
+    private final String SAVE_FILE_NAME = "./src/main/java/app/data/monobot_tasks.txt";
```
Would it be better if `SAVE_FILE_NAME` is static final?

Jensen188 commented on src/main/java/app/SaveHandler.java:
```diff
@@ -0,0 +1,64 @@
+    public ArrayList<Task> loadTasks() {
+        File saveFile = new File(SAVE_FILE_NAME);
+        Scanner sc = null;
+        try {
+            sc = new Scanner(saveFile);
+        } catch (FileNotFoundException e) {
+            return new ArrayList<>();
+        }
```
I personally didn't think of that! Good idea👍

Jensen188 commented on src/main/java/app/commands/CommandType.java:
```diff
@@ -0,0 +1,12 @@
+    DeleteTask,
+    PrintTasklist,
+    PrintFindTasklist,
+    Exit,
```
Nice tree LOL

Jensen188 commented on src/main/java/app/events/MonoBotEventSource.java:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+    
+    protected void invokeTaskAddedEvent(Task task, int numTasks) {
```
Perhaps change to `invokeAddTaskEvent` so that it aligns to your `invokeStartBotEvent` format?

Jensen188 commented on src/main/java/app/events/MonoBotEventSource.java:
```diff
@@ -0,0 +1,62 @@
+    protected void invokePrintTasklistEvent(final ArrayList<Task> tasklist) {
+        for (MonoBotEventListener listener : listeners) {
+            listener.onPrintTasklistEvent(tasklist);
+        }
```
Same suggestion as above.

Jensen188 commented on src/main/java/app/gui/GUIDialogueBox.java:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
```
Maybe `flipBox` is clearer than `flip`?

Jensen188 commented on src/main/java/app/monobot/MonoBotInputParser.java:
```diff
@@ -0,0 +1,192 @@
+            throw new app.exceptions.NumberFormatException();
+        }
+        return new TaskIndexCommand(CommandType.DeleteTask, idx);
+    }
```
Clear name for all the `processInput` methods! Clean code💯

Jensen188 commented on src/main/java/app/tasks/Task.java:
```diff
@@ -0,0 +1,73 @@
+     * Encodes the task in a format for save data
+     * @return Encoded format of task
+     */
+    public String encodeTask() {
```
Is it more clear if rename to `encodeTaskAsString`?

Jensen188 commented on src/main/java/app/utility/DateTime.java:
```diff
@@ -0,0 +1,52 @@
+     * Formats the date and time according to OUTPUT_FORMAT
+     * @return Formatted DateTime String
+     */
+    public String formatAsOutputString() {
+        return this.datetime.format(DateTime.getOutputFormatter());
+    }
+
+    /**
+     * Formats the date and time according to INPUT_FORMAT
+     * @return Formatted DateTime String
+     */
+    public String formatAsInputString() {
+        return this.datetime.format(DateTime.getInputFormatter());
+    }
+
+    private static DateTimeFormatter getOutputFormatter() {
+        if (outputFormatter == null) {
+            outputFormatter = DateTimeFormatter.ofPattern(OUTPUT_FORMAT);
+        }
+        return outputFormatter;
+    }
+
+    private static DateTimeFormatter getInputFormatter() {
+        if (inputFormatter == null) {
+            inputFormatter = DateTimeFormatter.ofPattern(INPUT_FORMAT);
+        }
+        return inputFormatter;
```
Clear naming of functionalities for methods!

Jensen188 reviewed (commented):
LGTM. Like the code design and readability! Some nits to be fixed. Learnt a lot from your code👍

monobeartae committed:
Add assertions

Add java assertions to MonoBot class

The code in MonoBot class assumes that the task indexes are positive.

This can cause errors down the line in future due to bugs

Let's add assertions to ensure that the index is indeed positive (>= 0) to prevent possible errors from our code in the future

monobeartae committed:
Check Code Quality

monobeartae committed:
Improve code quality further

monobeartae committed:
Use stream

monobeartae committed:
Merge pull request #1 from monobeartae/branch-A-CodeQuality

Improve Code Quality

monobeartae committed:
Update tasks data

Update tasks data

I wrote the full commit message for the PR and not the commit by mistake. It is not registered and not visible

A full message commit with tag A-FullCommitMessage is required for the sake of my grades

Thus, this is an irrelevant commit in order to register the tag and practice writing a message, and needs to be pushed

monobeartae committed:
Merge pull request #2 from monobeartae/branch-A-Streams

Use stream

monobeartae committed:
Merge pull request #3 from monobeartae/branch-A-Assertions

Add assertions

monobeartae committed:
Add Help Command (C-Help)

monobeartae committed:
Add Better GUI and Name Application

monobeartae committed:
Add More Error Handling

monobeartae committed:
Add Screenshot of App

monobeartae committed:
Modify find command

monobeartae committed:
Update User Guide

monobeartae committed:
Update README.md

monobeartae committed:
Update README.md

monobeartae committed:
Fix Save on App Close Bugs

monobeartae committed:
Merge branch 'master' of https://github.com/monobeartae/ip

monobeartae committed:
Fix Bug: Did not check tas index 0
